lib/libc/newlib/CMakeLists.txt

```
@@ -73,6 +73,9 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
 endif()
 zephyr_link_libraries(c)
 
+# newlib needs this to expose the full Zephyr C Library API
+zephyr_compile_definitions(_POSIX_C_SOURCE=200809L)
```

Is it possible to use this instead? That would change the definition to a local one rather than a global one (which would conflict in many places).
```suggestion
zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809L)
```

---

No, the whole point of this define is to ensure that newlib exposes the Zephyr C API as defined by Rule A.4 in the Coding Guidelines. Unless and until newlib itself is fixed to provide Zephyr compatibility, we can either

 1. Abandon Rule A.4 and assert that Zephyr only uses ISO C interfaces
 2. Define `_POSIX_C_SOURCE=200809L` globally
 3. Find some alternative mechanism to get newlib to expose the extended functions.

---

@keith-packard would it be an option to only expose by default the ISO C interface, but still allow those extra functions defined in Rule A.4 (thru individual waives in PRs that need to use them, where setting the POSIX_C_SOURCE macro is allowed in particular cases).

---

Are you saying that we *require* applications define `_POSIX_C_SOURCE=200809L` to gain access to the extended functions? Because if it's not required, then we're adding implicit dependencies on using only a C library with Zephyr compatibility (picolibc or the minimal C library).

That would certainly be easier for C library developers, but it seems like it will require careful auditing of changes in any file with such a definition to ensure that it doesn't accidentally start depending on additional POSIX functionality.

---

> Are you saying that we require ..

Yes

---

It seems like that would also require allowing core Zephyr files to contain such definitions; have tooling that prevents valid code from being merged will slow down development. Because once the definition is present in a source file, no future changes to that file will see the same review, hence allowing non-approved functions to leak in. Alternatively, more sophisticated tooling that detected the use of POSIX functions beyond the approved list.

---

You are right that in general that would be the case, and that would be very far from ideal.

Though today a quick check seems to show:
* there is nothing in the "zephyr kernel" using `strnlen()` (the only exception in A.4).
* for more peripheral code in the tree, there is a few uses of `strnlen()` and `strtok_r()` (rule A.5 exceptions).

But given that today, at least for the tests we are running, all those built with just these 
https://github.com/zephyrproject-rtos/zephyr/pull/67040/commits/123bdfb976689e7ea73aa7917d08dd9e1b3880bb
I wonder if it would be best to, by now, postpone this particular point (I would very much like to have @stephanosio on it), and not do this particular line change (or this other commit) in this PR; The rest of the PR seems fully uncontroversial and an overall improvement.

---

We can't remove instances of `#define _POSIX_C_SOURCE` in most places without somehow dealing with newlib's lack of Zephyr support. Otherwise, building code with newlib that uses `strnlen` will fail.